server/s3_copy_object_action.cc

```diff
@@ -0,0 +1,208 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#include "s3_copy_object_action.h"
+#include "s3_log.h"
+#include "s3_error_codes.h"
+#include "s3_common_utilities.h"
+#include "s3_motr_layout.h"
+#include "s3_uri_to_motr_oid.h"
+#include "s3_m0_uint128_helper.h"
+
+S3CopyObjectAction::S3CopyObjectAction(
+    std::shared_ptr<S3RequestObject> req, std::shared_ptr<MotrAPI> motr_api,
+    std::shared_ptr<S3BucketMetadataFactory> bucket_meta_factory,
+    std::shared_ptr<S3ObjectMetadataFactory> object_meta_factory,
+    std::shared_ptr<S3MotrWriterFactory> motrwriter_s3_factory,
+    std::shared_ptr<S3MotrReaderFactory> motrreader_s3_factory,
+    std::shared_ptr<S3MotrKVSWriterFactory> kv_writer_factory)
+    : S3ObjectAction(std::move(req), std::move(bucket_meta_factory),
+                     std::move(object_meta_factory)),
+      write_in_progress(false),
+      read_in_progress(false) {
+  s3_log(S3_LOG_DEBUG, request_id, "Constructor\n");
+  s3_log(S3_LOG_INFO, request_id,
+         "S3 API: CopyObject. Destination: [%s], Source: [%s]\n",
+         request->get_object_uri().c_str(),
+         request->get_copy_object_source().c_str());
+  s3_copy_action_state = S3CopyObjectActionState::empty;
+
+  old_object_oid = {0ULL, 0ULL};
+  old_layout_id = -1;
+  new_object_oid = {0ULL, 0ULL};
+
+  if (motr_api) {
+    s3_motr_api = std::move(motr_api);
+  } else {
+    s3_motr_api = std::make_shared<ConcreteMotrAPI>();
+  }
+  S3UriToMotrOID(s3_motr_api, request->get_object_uri().c_str(), request_id,
+                 &new_object_oid);
+  // Note valid value is set during create object
+  layout_id = -1;
+  tried_count = 0;
+  salt = "uri_salt_";
+
+  if (motrwriter_s3_factory) {
+    motr_writer_factory = std::move(motrwriter_s3_factory);
+  } else {
+    motr_writer_factory = std::make_shared<S3MotrWriterFactory>();
+  }
+  if (motrreader_s3_factory) {
+    motr_reader_factory = std::move(motrreader_s3_factory);
+  } else {
+    motr_reader_factory = std::make_shared<S3MotrReaderFactory>();
+  }
+  setup_steps();
+};
+
+void S3CopyObjectAction::setup_steps() {
+  s3_log(S3_LOG_DEBUG, request_id, "Setting up the action\n");
+  ACTION_TASK_ADD(S3CopyObjectAction::validate_copyobject_request, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::create_object, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::read_object, this);
```

Comment:
instead of read_object have copy_object() or initiate_data_copy() that will internally do read from source and then write to target

Reply:
Accepted.